src/utils.zig

```diff
@@ -0,0 +1,41 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+pub inline fn div_ceil(numerator: anytype, denominator: anytype) @TypeOf(numerator, denominator) {
+    comptime {
+        const T = @TypeOf(numerator, denominator);
+        switch (@typeInfo(T)) {
+            .Int => |int| assert(int.signedness == .unsigned),
+            .ComptimeInt => {
+                assert(numerator >= 0);
+                assert(denominator > 0);
+            },
+            else => @compileError("invalid type given to utils.div_ceil"),
+        }
+    }
+
+    // Only check the denominator at runtime when the it is *not* a comptime_int.
+    if (@as(std.builtin.TypeId, @typeInfo(@TypeOf(denominator))) == .Int) {
```

Comment:
Is the intention here to elide the assertion when it's a comptime int for performance? If so, I think we can collapse this and rather share across unsigned and comptime ints, because if the value is comptime known then the assertion should be elided anyway.

The reason I'm thinking this, is that for safety, it would be better to have a single code path where we check the denominator in one place only. This makes it impossible then to have any kind of semantic gap where type changes might mean the check is skipped.

@ifreund what do you think of this div_ceil as a whole? @sentientwaffle we were chatting today that it would be great to try and get this into the std lib, as some kind of divCeilAssert.

Reply:
> Is the intention here to elide the assertion when it's a comptime int for performance? If so, I think we can collapse this and rather share across unsigned and comptime ints, because if the value is comptime known then the assertion should be elided anyway.

Yes, that was the intent. I didn't know that it would be elided automatically when `denominator` is a comptime int.

---

Re:std lib: It would be nice if this use case was covered in the std lib, though I think it's important that the APIs be consistent. The existing `std.math.divCeil` takes a type as the first argument, rather than returning the peer-type-resolution of the numerator and denominator.

Is there a way to vary the return type based on whether or not the `denominator` is a comptime int? If so, I think we could get away with one function, since our use case always has a comptime-known denominator.

Reply:
> Is there a way to vary the return type based on whether or not the denominator is a comptime int?

Yes, there is! Check out how cool Zig's type system is 😁 

https://github.com/ziglang/zig/blob/master/lib/std/mem.zig#L2492-L2505

Reply:
Perfect! In that case, stdlib's `divCeil` could return a `!T` most of the time, but simply return a `T` when the denominator is comptime-known to be nonzero.

Reply:
> Re:std lib: It would be nice if this use case was covered in the std lib, though I think it's important that the APIs be consistent. The existing std.math.divCeil takes a type as the first argument, rather than returning the peer-type-resolution of the numerator and denominator.

@ifreund  and I were chatting yesterday again about `divCeil`. Let's rename `utils.zig` to `extensions.zig`. Then we call `ext.div_ceil()` in our source, i.e. with `ext` as the import namespace. Then we put everything we think we want from the std lib in `extensions.zig`, which will be our proposed std lib extensions at some point in the future, once we've had enough time to be sure they're well trodden.

Reply:
> Perfect! In that case, stdlib's divCeil could return a !T most of the time, but simply return a T when the denominator is comptime-known to be nonzero.

Yes, that's one way I was also thinking, but it might be confusing to people if they sometimes have to type `try` and sometimes not` and if it only works for comptime values.

Perhaps our `div_ceil` would be more of an internally asserting rather than error returning function completely, i.e. becoming `divCeilAssert` in the std lib.